napari/utils/key_bindings/dispatch.py

```diff
@@ -0,0 +1,353 @@
+    ----------
+    registry: NapariKeyBindingsRegistry
+        Registry providing the keymap to use for resolving key bindings.
+    context: Context
+        Mutable evented context with which to evaluate the key binding entries.
+    os: Optional[OperatingSystem]
+        Operating system with which to translate key bindings.
+
+    Attributes
+    ----------
+    dispatch: Signal(DispatchFlags, Optional[str])
+        Signal used to dispatch key binding related logic, containing flags for
+        how the dispatch should be done as well as the command, if found.
+    active_combo: int
+        Currently active key combo.
+    active_command: Optional[str]
+        Current command being processed.
+    is_prefix: bool
+        Whether the current active combo is a prefix for another key binding.
+    prefix: int
+        Current active prefix.
+    """
+
+    dispatch = Signal(DispatchFlags, Optional[str])  # type: ignore
+
+    def __init__(
+        self,
+        registry: NapariKeyBindingsRegistry,
+        context: Context,
+        os: Optional[OperatingSystem] = None,
+    ) -> None:
+        self.registry = registry
+        self.context = context
+        if os is None:
+            os = OperatingSystem.current()
+        self.os = os
+
+        self.active_combo: int = 0
+        self.active_command: Optional[str] = None
+        self.is_prefix: bool = False
+        self.prefix: int = 0
+
+        self._active_match_cache: Dict[int, Optional[KeyBindingEntry]] = {}
+        self._conflicts_cache: Dict[int, bool] = {}
+        self._active_keymap: Optional[Mapping[int, str]] = None
+
+        self.context.changed.connect(self._on_context_change)
+        self.registry.registered.connect(self._refresh_cache)
+        self.registry.unregistered.connect(self._refresh_cache)
+
+    def _on_context_change(self, changes: Set[str]):
+        logger.info('context change detected for keys %s', changes)
+        self._refresh_cache()
+        logger.debug('current context: %s', self.context)
+
+    def _refresh_cache(self):
+        logger.info('clearing cache')
+        self._active_match_cache.clear()
+        self._conflicts_cache.clear()
+        self._active_keymap = None
+
+    def find_active_match(self, key: int) -> Optional[KeyBindingEntry]:
+        """Find the active match for the key sequence provided.
+
+        Parameters
+        ----------
+        key: int
+            Key sequence to search.
+
+        Returns
+        -------
+        match: Optional[KeyBindingEntry]
+            Match, if found.
+        """
+        try:
+            match = self._active_match_cache[key]
+            logger.info('cached match found for %s: %s', key, match)
+        except KeyError:
+            logger.debug('cached match not found for %s', key)
+            entries = self.registry.keymap.get(key)
+            if not entries:
+                match = None
+            else:
+                match = find_active_match(entries, self.context)
```

Comment:
I guess it isn't likely that this will lead to an overflow since keymaps probably won't be too deep unless someone starts autogenerating them.